fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/ComparatorCursor.java

```diff
@@ -0,0 +1,346 @@
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.google.common.base.Suppliers;
+import com.google.protobuf.Message;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A cursor that implements a comparison of matching elements from a set of cursors all of whom are ordered compatibly.
+ * For each record returned by the sub-cursors, this cursor will ensure that they are the same (by using the comparison
+ * function). The cursor will log and continue if the records are not the same or if one of the sub-cursors ends
+ * prematurely.
+ * The cursor defines a "reference cursor" whose results are assumed to be correct. Values are compared to this
+ * cursor's results, and the reference cursor's results are the ones that are actually returned.
+ *
+ * @param <T> the type of elements returned by the cursor
+ */
+@API(API.Status.EXPERIMENTAL)
+public class ComparatorCursor<T> extends MergeCursor<T, T, KeyedMergeCursorState<T>> {
+    @Nonnull
+    private static final Logger logger = LoggerFactory.getLogger(ComparatorCursor.class);
+    @Nonnull
+    private static final Set<StoreTimer.Event> duringEvents = Collections.singleton(FDBStoreTimer.Events.QUERY_COMPARATOR);
+    @Nonnull
+    private static final Set<StoreTimer.Count> matchesCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_MATCH);
+    @Nonnull
+    private static final Set<StoreTimer.Count> mismatchesCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_MISMATCH);
+    @Nonnull
+    private static final Set<StoreTimer.Count> compareCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_COMPARED);
+
+    // The "reference plan" whose values are to be compared with the rest of the cursors, and whose values are actually returned
+    private final int referencePlanIndex;
+    @Nonnull
+    private final Supplier<String> planStringSupplier;
+    @Nonnull
+    private final Supplier<Integer> planHashSupplier;
+    private boolean errorLogged = false;
+
+    private ComparatorCursor(@Nonnull List<KeyedMergeCursorState<T>> cursorStates,
+                             @Nullable FDBStoreTimer timer,
+                             final int referencePlanIndex,
+                             @Nonnull final Supplier<String> planStringSupplier,
+                             @Nonnull final Supplier<Integer> planHashSupplier) {
+        super(cursorStates, timer);
+        this.referencePlanIndex = referencePlanIndex;
+        this.planStringSupplier = Suppliers.memoize(planStringSupplier::get);
+        this.planHashSupplier = Suppliers.memoize(planHashSupplier::get);
```

Comment:
I *think* the `::get` here should be extraneous because the arguments are already functions (or `Suppliers`), so they don't need to be lambda-ized

Reply:
Yeah, that was my gut feel too, but then:
`reason: no instance(s) of type variable(s) T exist so that Supplier<String> conforms to Supplier<T>`

Reply:
oof, that's unfortunate. I wonder if this is a bug in the compiler, or something, because it seems like it should conform... Alas

Reply:
I don't think the error message you paste applies to the the code here. What `T` is it talking about?

Out of interest I changed this to

```
this.planStringSupplier = planStringSupplier;
this.planHashSupplier = planHashSupplier;
```

and it builds fine. Please reconsider!

Reply:
Done.